beacon-chain/sync/data_columns_sampling.go

```diff
@@ -4,232 +4,271 @@ import (
-func (s *Service) sampleDataColumnsFromPeer(
-	pid peer.ID,
-	requestedColumns map[uint64]bool,
-	root [fieldparams.RootLength]byte,
-) (map[uint64]bool, error) {
-	// Build the data column identifiers.
-	dataColumnIdentifiers := make(types.DataColumnSidecarsByRootReq, 0, len(requestedColumns))
-	for index := range requestedColumns {
-		dataColumnIdentifiers = append(dataColumnIdentifiers, &eth.DataColumnIdentifier{
-			BlockRoot:   root[:],
-			ColumnIndex: index,
-		})
+	custodyColumnsCount := uint64(len(columns))
+	if peerdas.CanSelfReconstruct(custodyColumnsCount) {
+		log.WithFields(logrus.Fields{
+			"custodyColumnsCount": custodyColumnsCount,
+			"totalColumns":        params.BeaconConfig().NumberOfColumns,
+		}).Debug("The node custodies at least the half the data columns, no need to sample")
+		return
 	}
 
-	// Send the request.
-	roDataColumns, err := SendDataColumnSidecarByRoot(s.ctx, s.cfg.clock, s.cfg.p2p, pid, s.ctxMap, &dataColumnIdentifiers)
-	if err != nil {
-		return nil, errors.Wrap(err, "send data column sidecar by root")
+	// initialize missing columns.
+	d.missingColumns = make(map[uint64]bool)
+	for i := uint64(0); i < params.BeaconConfig().NumberOfColumns; i++ {
+		if exists := columns[i]; !exists {
+			d.missingColumns[i] = true
+		}
 	}
 
-	retrievedColumns := make(map[uint64]bool, len(roDataColumns))
+	// initialize peer info first.
+	d.refreshPeerInfo()
 
-	// Remove retrieved items from rootsByDataColumnIndex.
-	for _, roDataColumn := range roDataColumns {
-		retrievedColumn := roDataColumn.ColumnIndex
+	// periodically refresh peer info to keep peer <-> column mapping up to date.
+	async.RunEvery(ctx, PeerRefreshInterval, d.refreshPeerInfo)
 
-		actualRoot := roDataColumn.BlockRoot()
+	// start the sampling loop.
+	d.samplingRoutine(ctx)
+}
 
-		// Filter out columns with incorrect root.
-		if actualRoot != root {
-			// TODO: Should we decrease the peer score here?
-			log.WithFields(logrus.Fields{
-				"peerID":        pid,
-				"requestedRoot": fmt.Sprintf("%#x", root),
-				"actualRoot":    fmt.Sprintf("%#x", actualRoot),
-			}).Warning("Actual root does not match requested root")
+func (d *dataColumnSampler1D) samplingRoutine(ctx context.Context) {
+	stateCh := make(chan *feed.Event, 1)
+	stateSub := d.stateNotifier.StateFeed().Subscribe(stateCh)
+	defer stateSub.Unsubscribe()
 
-			continue
+	for {
+		select {
+		case evt := <-stateCh:
+			d.handleStateNotification(ctx, evt)
+		case err := <-stateSub.Err():
+			log.WithError(err).Error("DataColumnSampler1D subscription to state feed failed")
+		case <-ctx.Done():
+			log.Debug("Context canceled, exiting data column sampling loop.")
+			return
 		}
+	}
+}
 
-		// Filter out columns that were not requested.
-		if !requestedColumns[retrievedColumn] {
-			// TODO: Should we decrease the peer score here?
-			columnsToSampleList := sortedSliceFromMap(requestedColumns)
+// Refresh peer information.
+func (d *dataColumnSampler1D) refreshPeerInfo() {
```

Comment:
Could you, at the end of this function:
- Add a *debug* log with the content of the `columnFromPeer` dictionnary 
- Add a *warning* log with the columns index for which we have no peer at all